src/packet/types/key.rs

```diff
@@ -257,11 +261,137 @@ macro_rules! key {
+                            }
+                            PublicParams::Elgamal { p, g, y } => {
+                                packet.extend(mpi_to_vec(p));
+                                packet.extend(mpi_to_vec(g));
+                                packet.extend(mpi_to_vec(y));
+                            }
+                        }
+
+                        let mut length_buf: [u8; 2] = [0; 2];
+                        BigEndian::write_uint(&mut length_buf, packet.len() as u64, 2);
+
+                        let mut h = Hasher::new(MessageDigest::sha1()).unwrap();
+
+                        h.update(&[0x99]).unwrap();
+                        h.update(&length_buf).unwrap();
+                        h.update(&packet).unwrap();
+
+                        h.finish().unwrap().deref().to_vec()
+                    }
+
+                    KeyVersion::V2 | KeyVersion::V3 => {
+                        let mut h = Hasher::new(MessageDigest::md5()).unwrap();
+
+                        let mut packet = Vec::new();
+
+                        match &self.public_params {
+                            PublicParams::RSA { n, e } => {
+                                packet.extend(mpi_to_vec(n));
+                                packet.extend(mpi_to_vec(e));
+                            }
+                            PublicParams::DSA { p, q, g, y } => {
+                                packet.extend(mpi_to_vec(p));
+                                packet.extend(mpi_to_vec(q));
+                                packet.extend(mpi_to_vec(g));
+                                packet.extend(mpi_to_vec(y));
+                            }
+                            PublicParams::ECDSA { curve, p } => {
+                                packet.extend(curve.oid().iter().cloned());
+                                packet.extend(mpi_to_vec(p));
+                            }
+                            PublicParams::ECDH {
+                                curve,
+                                p,
+                                hash,
+                                alg_sym,
+                            } => {
+                                packet.extend(curve.oid().iter().cloned());
+                                packet.extend(mpi_to_vec(p));
+                                packet.push(*hash);
+                                packet.push(*alg_sym);
+                            }
+                            PublicParams::Elgamal { p, g, y } => {
+                                packet.extend(mpi_to_vec(p));
+                                packet.extend(mpi_to_vec(g));
+                                packet.extend(mpi_to_vec(y));
+                            }
+                        }
+
+                        h.update(&packet).unwrap();
+
+                        h.finish().unwrap().deref().to_vec()
+                    }
+                }
+            }
+
+            pub fn key_id(&self) -> Vec<u8> {
```

Comment:
given that this is not always supported I think returning `Option<Vec>` would be better

Reply:
What would result in returning None? I can't find much on the specification :/

Reply:
if it’s a v3 key with non rsa algorithm, though, instead of an empty vec